src/lib.rs

```diff
@@ -67,46 +67,134 @@ pub use attr::{AttributeValue, AttributeValueParts, Attributes};
+/// ```
+///
+/// Write to standard output with buffering ([`std::io::Stdout`] implements [`std::io::Write`]):
+///
+/// ```
+/// # use jotdown::Render;
+/// # let events = std::iter::empty();
+/// let mut out = std::io::BufWriter::new(std::io::stdout());
+/// let mut renderer = jotdown::html::Renderer::default();
+/// renderer.write(events, &mut out).unwrap();
+/// ```
 pub trait Render {
-    /// Push [`Event`]s to a unicode-accepting buffer or stream.
-    fn push<'s, I: Iterator<Item = Event<'s>>, W: fmt::Write>(
-        &self,
-        events: I,
-        out: W,
-    ) -> fmt::Result;
-
-    /// Write [`Event`]s to a byte sink, encoded as UTF-8.
+    /// Render a single event.
+    fn render_event<'s, W>(&mut self, e: &Event<'s>, out: W) -> std::fmt::Result
+    where
+        W: std::fmt::Write;
+
+    /// Render something before any events have been provided.
+    fn render_prologue<W>(&mut self, _out: W) -> std::fmt::Result
+    where
+        W: std::fmt::Write,
+    {
+        Ok(())
+    }
+
+    /// Render something after all events have been provided.
+    fn render_epilogue<W>(&mut self, _out: W) -> std::fmt::Result
+    where
+        W: std::fmt::Write,
+    {
+        Ok(())
+    }
+
+    /// Push owned [`Event`]s to a unicode-accepting buffer or stream.
+    fn push<'s, I, W>(&mut self, mut events: I, mut out: W) -> fmt::Result
+    where
+        I: Iterator<Item = Event<'s>>,
+        W: fmt::Write,
+    {
+        self.render_prologue(&mut out)?;
+        events.try_for_each(|e| self.render_event(&e, &mut out))?;
+        self.render_epilogue(&mut out)
+    }
+
+    /// Write owned [`Event`]s to a byte sink, encoded as UTF-8.
     ///
     /// NOTE: This performs many small writes, so IO writes should be buffered with e.g.
     /// [`std::io::BufWriter`].
-    fn write<'s, I: Iterator<Item = Event<'s>>, W: io::Write>(
-        &self,
-        events: I,
-        out: W,
-    ) -> io::Result<()> {
-        struct Adapter<T: io::Write> {
-            inner: T,
-            error: io::Result<()>,
-        }
+    fn write<'s, I, W>(&mut self, events: I, out: W) -> io::Result<()>
+    where
+        I: Iterator<Item = Event<'s>>,
+        W: io::Write,
+    {
+        let mut out = WriteAdapter {
+            inner: out,
+            error: Ok(()),
+        };
 
-        impl<T: io::Write> fmt::Write for Adapter<T> {
-            fn write_str(&mut self, s: &str) -> fmt::Result {
-                match self.inner.write_all(s.as_bytes()) {
-                    Ok(()) => Ok(()),
-                    Err(e) => {
-                        self.error = Err(e);
-                        Err(fmt::Error)
-                    }
-                }
-            }
+        match self.push(events, &mut out) {
+            Ok(()) => Ok(()),
+            Err(_) => match out.error {
+                Err(_) => out.error,
+                _ => Err(io::Error::new(io::ErrorKind::Other, "formatter error")),
+            },
```

Comment:
After writing this, I realise it could be simplified a bit, as:

```rust
self.push(events, &mut out).map_err(|_| match out.error {
    Err(e) => e,
    _ => io::Error::new(io::ErrorKind::Other, "formatter error"),
})
```

Reply:
fixed, good to minimize the duplicate code